Update native_functions.yaml to add DefaultBackend.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46127 Try remove catchAllKenrel_.
* #45939 Move catchAll to Math
* #46126 Add guideline about which dispatch keyword to use in native_functions.yaml.
* **#45938 Update native_functions.yaml to add DefaultBackend.**



Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

Commit:
Update native_functions.yaml to add DefaultBackend.

[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."




[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."




[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."




[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."




[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."


Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."


Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."


Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."


Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

[ghstack-poisoned]

ailzhang commented:
The `native_functions.yaml` changes is generated from @ljk53 's static analysis https://github.com/pytorch/pytorch/pull/44963 (huge thanks!) plus a few added for ops in `VariableTypeManual.cpp`.

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."


Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."


Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

[ghstack-poisoned]

ailzhang commented:
@bhosmer @ezyang May I get a review for this PR when you have time? ;) 
It's syncing derivatives.yaml's information to native_functions.yaml so there's a race condition between landing it and other PRs touching derivatives.yaml. Thanks!

ezyang commented:
Don't forget to write the PR desc ;) ;) ;)

ezyang commented on tools/codegen/gen.py:
```diff
@@ -897,7 +897,7 @@ def compute_registration_declarations(f: NativeFunction) -> str:
         'schema': f'aten::{f.func}',
         'dispatch': str(f.dispatch is not None),
-        'math': str(f.dispatch is not None and 'Math' in f.dispatch)
+        'default': str(f.dispatch is not None and ('Math' in f.dispatch or 'DefaultBackend' in f.dispatch))
```
Is there a paired XLA change? Or was math not being used yet.

ailzhang replied:
Yea there's a paired XLA change

bhosmer replied:
This is re-triggering my sense that `['Math', 'DefaultBackend']` is a semantically significant grouping and that we should define it somewhere (maybe not model.py, but somewhere), per the review comment in #45719. I don't know if it's urgent to do before landing this PR, but I'd like us to understand what common property it is that causes these two keys to show show up together, and ideally capture it in a definition.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -746,6 +746,13 @@ def gen_variable_type_shard(out, aten_declarations, template_path, suffix, heade
+        if declaration['name'] in MANUAL_AUTOGRAD_AND_TRACER or declaration['derivative']:
+            msg = "Did you add a formula for {}(or its functional variant) in derivatives.yaml? \
+                   If so please add a dispatch section for it with DefaultBackend in \
+                   native_functions.yaml.".format(declaration['name'])
```
nit: we got f-strings, use 'em!!

ezyang commented:
Don't bother addressing the review comments

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."


Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."


Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

[ghstack-poisoned]

Commit:
Update on "Update native_functions.yaml to add DefaultBackend."




Differential Revision: [D24165889](https://our.internmc.facebook.com/intern/diff/D24165889)

[ghstack-poisoned]
